File: enode/src/main/java/org/enodeframework/infrastructure/TypeUtils.java
package org.enodeframework.infrastructure;

import org.enodeframework.domain.IAggregateRepository;
import org.enodeframework.domain.IAggregateRoot;

import java.lang.reflect.Modifier;
import java.lang.reflect.ParameterizedType;

/**
 * @author anruence@gmail.com
 */
public class TypeUtils {
    public static boolean isAggregateRoot(Class type) {
        return !Modifier.isAbstract(type.getModifiers()) && IAggregateRoot.class.isAssignableFrom(type);
    }

    public static boolean isAggregateRepositoryType(Class type) {
        return type != null && !Modifier.isAbstract(type.getModifiers()) && IAggregateRepository.class.isAssignableFrom(type);
    }

    public static Class getGenericType(Class clazz) {
        return (Class) ((ParameterizedType) clazz.getGenericSuperclass()).getActualTypeArguments()[0];
    }
}

File: enode/src/main/java/org/enodeframework/infrastructure/TypeUtils.kt
package org.enodeframework.infrastructure

import org.enodeframework.domain.AggregateRepository
import org.enodeframework.domain.AggregateRoot
import java.lang.reflect.Modifier
import java.lang.reflect.ParameterizedType

/**
 * @author anruence@gmail.com
 */
object TypeUtils {
    fun isAggregateRoot(type: Class<*>): Boolean {
        return !Modifier.isAbstract(type.modifiers) && AggregateRoot::class.java.isAssignableFrom(type)
    }

    fun isAggregateRepositoryType(type: Class<*>): Boolean {
        return !Modifier.isAbstract(type.modifiers) && AggregateRepository::class.java.isAssignableFrom(type)
    }

    fun getGenericType(clazz: Class<*>): Class<*> {
        return (clazz.genericSuperclass as ParameterizedType).actualTypeArguments[0] as Class<*>
    }
}